src/utility/validation/AuthenticatorValidator.ts

```diff
@@ -0,0 +1,142 @@
+		allowlistedAuthenticators: Array<Aaid>
+	): Array<Authenticator> {
+		return AuthenticatorValidatorImpl.allowedAuthenticators(
+			context,
+			allowlistedAuthenticators
+		).filter((authenticator) => {
+			// Do not display:
+			//   - non-registered authenticators
+			//   - not hardware supported authenticators
+			return (
+				authenticator.isSupportedByHardware &&
+				authenticator.registration.isRegistered(context.account.username)
+			);
+		});
+	}
+}
+
+export class AuthenticatorValidatorImpl extends AuthenticatorValidator {
+	/**
+	 * Filters out non-allowlisted authenticators.
+	 * @param context The context holding the accounts to validate.
+	 * @param allowlistedAuthenticators The array holding the allowlisted authenticators.
+	 * @returns The array of allowed authenticators.
+	 */
+	static allowedAuthenticators(
+		context: AuthenticatorSelectionContext,
+		allowlistedAuthenticators: Array<Aaid>
+	): Array<Authenticator> {
+		return context.authenticators.filter((authenticator) => {
+			const authenticatorAaid = AuthenticatorUtils.getAaidFromRawValue(authenticator.aaid);
+			if (authenticatorAaid === undefined) {
+				return false;
+			}
+			return allowlistedAuthenticators.some((allowlistedAuthenticator) => {
+				console.log('');
+				return allowlistedAuthenticator.rawValue() === authenticatorAaid.rawValue();
+			});
+		});
+	}
+
+	static async filterAndroidFingerprintIfNecessary(
+		context: AuthenticatorSelectionContext,
+		authenticator: Authenticator
+	): Promise<boolean> {
+		if (Platform.OS === 'android') {
+			const biometricsRegistered = context.authenticators.filter((contextAuthenticator) => {
+				return (
+					contextAuthenticator.aaid === Aaid.BIOMETRIC.rawValue() &&
+					contextAuthenticator.registration.isRegistered(context.account.username)
+				);
+			});
+
+			const biometricsAvailable: Authenticator[] = [];
+			const fingerprintAvailable: Authenticator[] = [];
+			for (const contextAuthenticator of context.authenticators) {
+				const isBiometricAvailable =
+					contextAuthenticator.aaid === Aaid.BIOMETRIC.rawValue() &&
+					contextAuthenticator.isSupportedByHardware &&
+					(await context.isPolicyCompliant(contextAuthenticator.aaid));
+				if (isBiometricAvailable) {
+					biometricsAvailable.push(contextAuthenticator);
+				}
+
+				const isFingerprintAvailable =
+					contextAuthenticator.aaid === Aaid.FINGERPRINT.rawValue() &&
+					contextAuthenticator.isSupportedByHardware &&
+					(await context.isPolicyCompliant(contextAuthenticator.aaid));
+				if (isFingerprintAvailable) {
+					fingerprintAvailable.push(contextAuthenticator);
+				}
+			}
+
+			// If biometric can be registered (or is already registered), or if we
+			// cannot register fingerprint, do not propose to register fingerprint
+			// (we favor biometric over fingerprint).
+			if (
+				(biometricsRegistered.length > 0 ||
+					biometricsAvailable.length > 0 ||
+					fingerprintAvailable.length === 0) &&
```

Comment:
🔧 I know it's _TypeScript_ but the logic here is so hard to read and not just because of the language. There got to be a way to not need checking _Arrays_ (that are named like _booleans_) for length to see whether certain conditions are true or not. Can't we for example filter authenticators from the context similarly to when `biometricsRegistered` is assigned, instead of a for-loop? Also, can't we assign whether the filtered Array is empty or not to `biometricsRegistered` so it's a `boolean`?

Reply:
You are perfectly right, tried to keep it aligned with the Flutter version, but during the refactor I could not stop at the aligned version, needed to 'simplify' it a bit further.

Reply:
⛏️ Talking about alignment with Flutter, I would consider using the `some` method (see [tutorial](https://www.geeksforgeeks.org/typescript-array-some-method/ it's kind of similar to the any that is used in the Flutter implementation) - I think it would be even more readable, but I'm also okay with how you changed the implementation (though two small _nitpicks_ is that I would have used a forEach and contextAuthenticator is a weird name, even the one letter a as it was named in the _Flutter_ implementation is better IMHO but I would really not want to insist on these).

So I would leave it up to you whether you would like to try aligning it with _Flutter_ more (e.g. with the `some` method) but I will approve already.

Reply:
IMO there is no point to talk about the initial version now. Ofc I've used `some` already 😉, in the current version it does not sense to use it, as currently the idea was to only iterate over the authenticators only once, therefore the 3 if statements. Otherwise with `some` we'd need to iterate over the authenticators more.

Reply:
Yes, sure that's a trade-off that @tamas-toth went with in the Flutter implementation and favours readability over performance and I tend to agree that the performance optimization here worth less, but as I said, I'm okay with the current one and left it up to you which one you choose!

Reply:
IMO in the current solution, even that we do not use `some`, there is no tradeoff in terms of speed, this is the fastest solution. We only have to loop over the authenticator array once all-together in the whole function.
If needed, we could have a huddle about this later on somewhere.